### ISSUE TITLE
Samples during training

### ISSUE THREAD

**Issue by Oruli** [tdrussell/diffusion-pipe#331]
Is it possible to implement this? Most other trainers provide the feature and it's very helpful to know if training is going in the right direction or not. I know you aren't providing a UI but locally written samples would be good enough

**Comment by Oruli** [comment]
Just to add, if samples isn't the way, what is everyone else doing here to monitor their trains? I'm currently having to copy a 350mb WAN lora each epoch into comfy to test it, but A/B testing this way is ridiculously tedious.

**Comment by tdrussell** [comment]
My personal opinion is that sampling during training greatly complicates things and adds a large maintenance burden, for little benefit.

Each model needs to be changed to add inference support. Some models use different training objectives, and require different samplers. Inference has interactions with pipeline parallelism, and block swapping, which complicate the implementation. For video models like Wan, inference can be very slow unless you use something like the lightx2v lora for low step inferencing. Basically inference is its own can of worms, and I would rather just have this repo solely focus on the training side.

Here is how I think about evaluating models.

If you have a relatively small dataset (<100 images):
Just train it for like 100 epochs with a relatively conservative learning rate, saving every 10 epoch or so. It's a small dataset so this doesn't take that long. Now you have 10 checkpoints you can test using whatever inference program you prefer.

If you have a large dataset (think thousands of images, or hundreds of videos):
You should be using a validation set, and measuring stabilized validation loss. This will give you a very good idea of how your model is learning. Minimum validation loss is not the only thing that matters, but I have found it highly correlated with quality and flexibility of the trained model. Also, the training script supports resuming from checkpoint, so for very long training runs on large datasets, you can still kill the run, test models manually, then resume.

**Comment by Oruli** [comment]
> My personal opinion is that sampling during training greatly complicates things and adds a large maintenance burden, for little benefit.
> 
> Each model needs to be changed to add inference support. Some models use different training objectives, and require different samplers. Inference has interactions with pipeline parallelism, and block swapping, which complicate the implementation. For video models like Wan, inference can be very slow unless you use something like the lightx2v lora for low step inferencing. Basically inference is its own can of worms, and I would rather just have this repo solely focus on the training side.
> 
> Here is how I think about evaluating models.
> 
> If you have a relatively small dataset (<100 images): Just train it for like 100 epochs with a relatively conservative learning rate, saving every 10 epoch or so. It's a small dataset so this doesn't take that long. Now you have 10 checkpoints you can test using whatever inference program you prefer.
> 
> If you have a large dataset (think thousands of images, or hundreds of videos): You should be using a validation set, and measuring stabilized validation loss. This will give you a very good idea of how your model is learning. Minimum validation loss is not the only thing that matters, but I have found it highly correlated with quality and flexibility of the trained model. Also, the training script supports resuming from checkpoint, so for very long training runs on large datasets, you can still kill the run, test models manually, then resume.

Oh I agree video samples would be ridiculous, but not images, at training res like 512/1024 they would take a second or so. It would give a great indication even for Wan just sampling a frame.

I was using Fluxgym before this to train flux, it generated 5 samples every X epochs (configurable) in like 15 seconds or something. It takes me far far longer to copy of 3GB of loras (10 epochs), set them all up manually switching between them in comfy and generating images with 5 different prompts.

It's your code, but I can't agree that the two methods are even slightly comparable. I understand the maintenance implications but once added for flux, wan and maybe another 1 or 2 popular models they are there forever and hugely improve on user experience.
